src/vmm/src/config.rs

```diff
@@ -1,13 +1,87 @@
 // Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 // SPDX-License-Identifier: Apache-2.0 OR BSD-3-Clause
 
+use std::collections::HashMap;
 use std::convert::TryFrom;
 use std::fmt;
+use std::num;
 use std::path::PathBuf;
 use std::result;
+use std::str::FromStr;
 
 use super::{DEFAULT_HIGH_RAM_START, DEFAULT_KERNEL_CMDLINE};
 
+fn cfg_args_to_map(input: String) -> HashMap<String, String> {
+    input
+        .split(',')
+        .filter(|tok| !tok.is_empty())
+        .map(|tok| {
+            let mut iter = tok.splitn(2, '=');
+            let param_name = iter.next().unwrap();
+            let value = iter.next().unwrap_or("").to_string();
+            (param_name.to_lowercase(), value)
+        })
+        .collect()
+}
+
+/// Parse a single argument.
+///
+/// This function is currently a thing wrapper around the FromStr function.
+/// # Arguments
+///
+/// * `params` - An hashmap resulting from `cfg_param_parse` call.
+/// * `param_name` - The name of the parameter to parse.
+/// * `default` - A default value to use if this parameter was not supplied.
+fn cfg_param_parse<T: FromStr>(
+    params: &mut HashMap<String, String>,
+    param_name: &'static str,
+    default: Option<T>,
```

Comment:
I believe the default parameter is better handled outside of this function, because this function just parses parameters, and retrieves the value if it exists. The defaults are dependent on the caller, and should be handled in the calling function.

The function would need to return a `Result<Option<T>, CfgArgParseError>`, and the caller would need to provide a default as follows:
```rust
cfg_param_parse(
      tokens,
      "mem_size_mib",
) .map_err(ConversionError::new_memory)?.unwrap_or(256);
```

Reply:
Yes, this is also way better. I was also thinking - perhaps there's a more readable way to do this, but couldn't think of one. This is neat!

Reply:
Thanks for this suggestion(s)! I initially thought about it in this way to further reduce code duplication. Your proposal works well for arguments which have the default parameter (like this one). For args which don't have a default it would need to build a new Error with a "missing parameter" error. 

This is an example for `path` argument of Kernel config:
```
        let path = arg_parser
            .value_of("path")
            .map_err(ConversionError::new_kernel)?
            .ok_or_else(|| ConversionError::new_memory("Missing parameter: path"))?;
```
When querying a missing value from the hashmap, the return value is "Optional" only if we have a default value for it. Otherwise, it should be a Result because missing from the hashmap is an Error. 

* If the additional overhead is ok, I can proceed with that. 
* Otherwise, what if argparse provides another function similar which returns an Error if the parameter is missing, thus with type `Result<T, CfgParseError>`? It can also live as a function outside the ArgParser struct if we don't want it there. I would call it something like `required_value_of()`.

Reply:
> When querying a missing value from the hashmap, the return value is "Optional" only if we have a default value for it. Otherwise, it should be a Result because missing from the hashmap is an Error.
> 

I believe it's the caller responsibility to decide weather a `None` is an `Error`. We basically only return errors when there was a value to begin with. From a separation of concerns point of view, I believe it is better to have the occasional duplication if needed.

Another option would be to have a function that treats this scenario in ArgParser. I was thinking of something that just does this for String, as follows:
```rust
    fn value_of_required<T: FromStr>(&mut self, param: &'static str) -> Result<String, CfgArgParseError> {
        self.value_of::<String>(param)?.ok_or(Err(CfgArgParseError::MissingRequired(param.to_string()))
    }
```

And then when parsing the kernel we would have:
```rust
let path = tokens.value_of_required("path")?;
if path.is_empty() {
    .....
}
```
There would still be a bit of duplication, but I would find it acceptable either way. I believe separating concerns is more important than a few duplicated lines.